p4include/psa.p4

```diff
@@ -0,0 +1,672 @@
+/// \TODO: should this be removed in favor of letting the extern
+/// return whatever distribution is supported by the target?
+enum RandomDistribution {
+  PRNG,
+  Binomial,
+  Poisson
+}
+
+extern Random<T> {
+  Random(RandomDistribution dist, T min, T max);
+  T read();
+
+  /*
+  @ControlPlaneAPI
+  {
+    void reset();
+    void setSeed(in T seed);
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ *  Action profiles are used as table implementation attributes
+ *
+ * Action profiles implement a mechanism to populate table entries
+ * with actions and action data. The only data plane operation
+ * required is to instantiate this extern. When the control plane adds
+ * entries (members) into the extern, they are essentially populating
+ * the corresponding table entries.
+ */
+extern ActionProfile {
+  /// Construct an action profile of 'size' entries
+  ActionProfile(bit<32> size);
+
+  /*
+  @ControlPlaneAPI
+  {
+     entry_handle add_member    (action_ref, action_data);
+     void         delete_member (entry_handle);
+     entry_handle modify_member (entry_handle, action_ref, action_data);
+  }
+  */
+}
+
+/**
+ *  Action selectors are used as table implementation attributes
+ *
+ * Action selectors implement another mechanism to populate table
+ * entries with actions and action data. They are similar to action
+ * profiles, with additional support to define groups of
+ * entries. Action selectors require a hash algorithm to select
+ * members in a group. The only data plane operation required is to
+ * instantiate this extern. When the control plane adds entries
+ * (members) into the extern, they are essentially populating the
+ * corresponding table entries.
+ */
+extern ActionSelector {
+  /// Construct an action selector of 'size' entries
+  /// @param algo hash algorithm to select a member in a group
+  /// @param size number of entries in the action selector
+  /// @param outputWidth size of the key
+  ActionSelector(HashAlgorithm algo, bit<32> size, bit<32> outputWidth);
+
+  /*
+  @ControlPlaneAPI
+  {
+     entry_handle add_member        (action_ref, action_data);
+     void         delete_member     (entry_handle);
+     entry_handle modify_member     (entry_handle, action_ref, action_data);
+     group_handle create_group      ();
+     void         delete_group      (group_handle);
+     void         add_to_group      (group_handle, entry_handle);
+     void         delete_from_group (group_handle, entry_handle);
+  }
+  */
+}
+
+
+
+/* --------------------------------------------------------------------- */
+/// \TODO: is generating a new packet and sending it to the stream or is
+/// it adding a header to the current packet and sending it to the
+/// stream (copying or redirecting).
+extern Digest<T> {
+  Digest(PortId_t receiver); /// define a digest stream to receiver
+  void emit(in T data);      /// emit data into the stream
```

Comment:
This is a bmv2 specific question. The current compiler always copies the data to a temporary for digest.